### ISSUE TITLE
[Flang] Missing diagnostic on mixed function and subroutine with the same generic name.

### ISSUE THREAD

**Issue by DanielCChen** [llvm/llvm-project#128757]
Consider the following code:
```
module m
  type dt
  contains
    procedure :: foo
    procedure :: bar
    generic :: gen => foo, bar  ! Error: foo is function, but bar is a sub
  end type
contains
  integer function foo(a)
    class(dt) a
    foo = 3
  end function

  subroutine bar(a)
    class(dt) a
  end subroutine
end module
```

`foo` and `bar` shall not be mixed into the same generic name `gen`.

**Comment by klausler** [comment]
This is an intentional extension.  It's never ambiguous -- function references and subroutine calls are syntactically distinct -- and it prevents spurious errors from arising when generics from multiple modules are combined by USE association.

**Comment by DanielCChen** [comment]
Found it in the `Extensions.md`
```
f18 does not enforce a blanket prohibition against generic interfaces containing a mixture of functions and subroutines. We allow both to appear, unlike several other Fortran compilers. This is especially desirable when two generics of the same name are combined due to USE association and the mixture may be inadvertent.
```
